File: day11/solution part 1.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <fstream>
#include <cstdint>

std::vector<uint64_t> update(std::vector<uint64_t>& stones){
     std::vector<uint64_t> updated_stones;
    for (uint64_t& stone: stones){
        if (stone == 0) updated_stones.push_back(1);
        else if (std::to_string(stone).length() %2 == 0){
            std::string stone_string = std::to_string(stone);
            int mid = stone_string.length()/2;

            uint64_t lhs = static_cast<uint64_t>(std::stoull(stone_string.substr(0, mid)));
            uint64_t rhs = static_cast<uint64_t>(std::stoull(stone_string.substr(mid)));

            updated_stones.push_back(lhs);
            updated_stones.push_back(rhs);
        }
        else updated_stones.push_back(stone * 2024);
    }
    return updated_stones;
}

int main() {
    std::ifstream file("data.txt"); std::string line, stone; std::vector<uint64_t> stones;
    std::getline(file, line); 
    file.close();

    std::stringstream ss(line);
    while (ss >> stone){
        stones.push_back(static_cast<uint64_t>(stoi(stone)));
    }

    for (int i{}; i < 25; ++i){
        stones = update(stones);
    }
    std::cout << stones.size();
    return 0;
}

File: day11/solution part 1.py


stones=[]
with open("data.txt", 'r') as file:
    stones = list(map(int, file.read().split()))

def update(stones: list[int]) -> list[int]:
    updated_stones = []
    for stone in stones:
        if stone == 0:
            updated_stones.append(1)
        elif len(str(stone)) %2 == 0:
            lhs, rhs = str(stone)[:len(str(stone))//2], str(stone)[len(str(stone))//2:]
            updated_stones.append(int(lhs))
            updated_stones.append(int(rhs))
        else:
            updated_stones.append(stone * 2024)
    return updated_stones

for i in range(25):
    stones = update(stones)

print(len(stones))